addons/sale_loyalty/models/sale_order.py

```diff
@@ -156,20 +156,46 @@ def _get_reward_values_product(self, reward, coupon, product=None, **kwargs):
             'tax_id': [(Command.CLEAR, 0, 0)] + [(Command.LINK, tax.id, False) for tax in taxes]
         }]
 
-    def _discountable_order(self, reward):
+    def _discountable_order(self, reward, is_reward_mandatory=True):
         """
-        Returns the discountable and discountable_per_tax for a discount that applies to the whole order
+        Computes `discountable` amount before applying the `reward` to the entire order
+        and maps taxes to its corresponding discountable amount in `discountable_per_tax`.
+
+        This method computes the total amount eligible for discount across order lines,
+        excluding those specified by `_get_no_effect_on_threshold_lines`(if not a payment program).
+        It also deducts fixed taxes from the `line.price_total` to ensure they are not discounted.
+
+        The method also maps non-fixed `taxes` to their total `amount_untaxed`,
+        providing a basis for detailed discount application by tax.
+
+        :param loyalty.reward reward: The reward to be applied, not yet considered and
+            must be applicable at the order level.
+
+        :param boolean is_reward_mandatory: If True, ensures that the discount will be
+            applied to the order.
+
+        :return: A tuple with the first element being the total discountable amount of the order,
+            and the second a dictionary mapping each non-fixed tax to its corresponding total
+            untaxed amount of the eligible order lines.
+        :rtype: tuple(float, dict(account.tax: float))
         """
         self.ensure_one()
-        assert reward.discount_applicability == 'order'
+        assert (not is_reward_mandatory and not reward) or reward.discount_applicability == 'order'
 
         discountable = 0
         discountable_per_tax = defaultdict(int)
-        lines = self.order_line if reward.program_id.is_payment_program else (self.order_line - self._get_no_effect_on_threshold_lines())
+        lines = self.order_line
+
+        if reward and not reward.program_id.is_payment_program:
+            lines -= self._get_no_effect_on_threshold_lines()
```

Comment:
I don't understand why this measure is only applied if it's not a payment program.